File: hashrename.js
#!/usr/bin/env node

/*
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

var fs = require('fs-extra'); //https://www.npmjs.com/package/node-fs-extra
var path = require('path');
var crypto = require('crypto')

//Global variable for recursion.
var recursive = false; //By default, we are not recursive.

function PrintHelp() {
	console.log("Usage: " + __filename + " [options] <dir(s)> <file(s)>\n"
		+ "	-R     | Recursively navigate directories.\n"
		+ "	--help | Print this help.\n\n"
		+ "Example: " + __filename + " /user/junk /mnt/stuff/things.txt\n");
	process.exit(-1);
}

if(process.argv.length <= 2) {
	PrintHelp();
}

function ProcessFile(full_file) {
	//Read the file and compute a hash.
	fs.createReadStream(full_file).pipe(crypto.createHash('sha256').setEncoding('hex')).on('finish', function () {
		var local_hash = this.read();
		var BaseFileName = path.basename(full_file); //Basename

		//Function to create our hashed path to move as a single variable.
		var FullHashPath = () => {
			//We consider everything past the first dot to be an extension.
			var GetFirstDot = (local_index) => {
				return BaseFileName.slice(local_index);
			}
			
			/*
			Parts:
			- Fully resolved dir name
			- Hash
			- Ext of the original filename.
			*/
			
			return path.join(path.resolve(path.dirname(full_file)), local_hash +  GetFirstDot(BaseFileName.indexOf('.')));
		}

		//Ensure we aren't trying to rename a file to itself.
		if(full_file != FullHashPath()) {
			fs.move(full_file, FullHashPath(), {clobber: true}, function (err2) {
				if(err2) { console.error(err2); }
							
				console.log('Successfully moved: ' + full_file + ' -> ' + FullHashPath());
			});
		}
	});
}

function ProcessPathArgs(current_path, basedir) {
	var IsRecursiveValid = () => {
		return(basedir || recursive);
	}

	//Function to check the details of the arguement provided
	fs.lstat(current_path).then(stats => {
		//Take this logic if the path is determined to be a file.
		if(stats.isFile()) {
			ProcessFile(current_path);
		}
		
		//Check if recursion is valid.
		else if(stats.isDirectory() && IsRecursiveValid()) {
			//Read file list into array.
			fs.readdir(current_path).then(files => {
				//Feed every file into the ProcessFile() function.
				files.forEach(file => {
					ProcessPathArgs(path.join(current_path,file),false);
				});
			}).catch(err => {
				console.error(err);
			});
		} else {
			console.error(current_path + ' is not a file or directory.');
		}
	}).catch(err => {
		console.error(err);
	});
}

function ProcessArgs() {
	var args = new Set(process.argv.slice(2));
	
	//If the help command was given at all, we just do help and quit.
	if(args.has('--help')) {
		PrintHelp();
	}

	//Set variable for dry run.
	if(args.has('-R')) {
		recursive = true;
		args.delete('-R')
	}

	args.forEach(current_arg => {
		ProcessPathArgs(path.resolve(current_arg),true);
	});
}

ProcessArgs();

File: hashrename.cpp
/*
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <iostream>
#include <filesystem>
#include <functional>
#include <cstring>
#include <cstdio>
#include <openssl/sha.h>
#include <libioapp.hpp>

//Globals
bool recursive = false;

std::string SHA256FromFile(const std::filesystem::path &p)
{
	const std::string &filename = p.string();

	unsigned char sh[SHA256_DIGEST_LENGTH];
	int bytes;
	FILE *shaFile = fopen (filename.c_str(), "rb");

	SHA256_CTX shContext;
	unsigned char data[1024];

	if(shaFile == NULL)
	{
		std::cerr << filename << " cannot be opened to hash." << std::endl;
		return "FAIL";
	}

	SHA256_Init(&shContext);

	while((bytes = fread(data, 1, 1024, shaFile)) != 0)
	{
		SHA256_Update(&shContext, data, bytes);
	}

	SHA256_Final(sh, &shContext);
	std::fclose(shaFile);

	char SHA256_return_string[65];
	for(int a = 0; a < SHA256_DIGEST_LENGTH; a++)
	{
		sprintf(SHA256_return_string + (a * 2), "%02x", sh[a]);
	}

	return SHA256_return_string;
}

int PrintHelp(void)
{
	std::cerr <<
	"Usage: hashrename [options] <dir(s)> <file(s)>\n" <<
	"	-R     | Recursively navigate directories.\n" <<
	"	--help | Print this help.\n\n" <<
	"Example: hashrename /user/junk /mnt/stuff/things.txt\n";
	return 1;
}

int main(int argc, char* argv[])
{
	ioapp::ProcessArgs hashrename_ProcessArgs(argc,argv,true,1);

	for(const std::string &f : hashrename_ProcessArgs.Flags)
	{
		if(f == "--help")
		{
			return PrintHelp();
		}

		if(f == "-R")
		{
			recursive = true;
		}
	}

	for(const std::string &arg : hashrename_ProcessArgs.Paths)
	{
		ioapp::ioapp(std::filesystem::path(arg), [&](std::filesystem::path p) {
			if(std::filesystem::is_regular_file(p))
			{
				std::string GetExt = "";
				if(p.has_extension())
				{
					GetExt = p.extension().string();
				}

				std::string new_file = p.parent_path().string() + '/' + SHA256FromFile(p) + p.extension().string();
				//We try two methos of moving a file, and crash if they fail.
				try
				{
					std::filesystem::rename(p, new_file);
				}

				catch(const std::exception& e)
				{
					try
					{
						std::filesystem::copy(p, new_file);
						std::filesystem::remove(p);
					}

					catch(const std::exception& ex)
					{
						std::cerr << ex.what() << '\n' << "Failed to move " << p.string() << std::endl;
						exit(1);
					}
				}

				std::cout << "Move: " << p.string() << " -> " << new_file << std::endl;
			}
			
			
		}, recursive);
	}

	return 0;
}